### ISSUE TITLE
Optimize Queries

### ISSUE THREAD

**Issue by btecu** [json-api-dotnet/JsonApiDotNetCore#404]
There are a few improvements that can be done to the auto generated endpoints:
 - GET endpoints could have `.AsNoTracking()` added to the query
 - DELETE endpoints could do `.Attach()` -> `.Remove()` instead of `.Get()` -> `.Remove()` to avoid making two calls
 - UPDATE could possibly be improved similarly to DELETE

**Comment by jaredcnance** [comment]
Very good points!

#221 was opened to address your first point and was implemented in #225. However, I closed the PR because I didn't have cycles at that time to complete it. From the PR:

> This is on hold until I can evaluate the scope of the breaking changes. Not only are there API implications but also impacts on possible extensibility scenarios. By setting the query as NoTracking, acceptance tests using the same context may be impacted.
> Need to put together documentation on these breaking changes and how users should handle acceptance testing that uses the same DbContext instance.
>
> **Update**
> I think the best way to move forward is to allow a global options flag to be set in Startup that will enable AsNoTracking. It will need to default to false to avoid breaking changes and will allow tests to enable tracking for all queries.
> There may also be impact to #238

In retrospect, based on experience with other projects (including [DotnetTestAbstractions](https://github.com/jaredcnance/DotnetTestAbstractions)), I think tests should not share the same `DbContext` (in most cases -- but there is definitely a notable performance benefit) and should instead share the underlying connection and transaction. So, it would be less of a concern. However, it should still be configurable at first and then default in a later major version.

**Comment by btecu** [comment]
How about having `GetRawAsync(bool isReadOnly)` and be called from the other endpoints?

Somewhat related, better naming could be nicer. Looking at `Get()` and `GetAsync()` is not obvious the difference. Both are `async` but the first returns a collection while the second one single record.

**Comment by jaredcnance** [comment]
I _think_ your reference to GetRawAsync is actually the intent of Get. 

Get is not async. It doesn’t actually run a query since it returns IQueryable. It serves as the foundation for (most) other requests.

**Comment by btecu** [comment]
Right, it could be called `GetRaw(bool isReadOnly)`.

**Comment by jaredcnance** [comment]
I'm open to discussion on the name change. For me, "GetRaw" doesn't actually clear the situation up any and begs additional questions. Perhaps just `GetQueryable`?

The rollout might look like:

```c#
// 1st release
IQueryable<TEntity> Get() => GetRaw(isReadOnly: false);
IQueryable<TEntity> GetRaw(bool isReadOnly = true);

// 2nd release
[Obsolete("Use GetRaw instead")]
IQueryable<TEntity> Get()  => GetRaw(isReadOnly: false);
IQueryable<TEntity> GetRaw(bool isReadOnly = true);

// 3rd (major) release
IQueryable<TEntity> GetRaw(bool isReadOnly = true);
```

The only thing here is that it makes global configuration more difficult which is what is needed for some testing scenarios (e.g. in `TestStartup`, you would set `UseAsNoTracking = false` if you're sharing the DbContext with your test). 

Another option that would address this is:

```c#
IQueryable<TEntity> GetRaw(bool? isReadOnly = null);
```

In the `null` or unspecified case, the global default would be used. This supports global and per-resource configuration.

**Comment by btecu** [comment]
I like `GetQueryable`.

Since version 3 is beta, does it have to go through all those releases?

**Comment by jaredcnance** [comment]
I think it should. Because if we skip straight to the breaking change, then developers will upgrade to 3.0.0 with no prior warning and may have to change a significant amount of their code base (albeit find and replace should be pretty straightforward). I prefer to deprecate first when possible. Step 1 and 2 can probably be combined though. And step 3 should actually keep the Obsolete attribute and simply set error: true

**Comment by wisepotato** [comment]
@maurei  status on this?

**Comment by bart-degreed** [comment]
Closing this, because:
- AsNoTracking is tracked at #221
- Delete-without-fetch is implemented by #792
- Update-without-fetch makes it impossible to track side effects, which is required by the json:api spec.
